# Aula 06: Estruturas de Repetição - Resolução e Lista de Exercícios

Agora que já sabemos como usar o `while` e o `for`, chegou a hora de praticarmos a construção de laços com exercícios selecionados.

## Exemplo Resolvido 1: Tabuada

**O problema:** Crie um programa que peça um número ao usuário e imprima a tabuada desse número de 1 a 10 usando o `for`.

In [ ]:
numero = int(input("Você quer a tabuada de qual número? "))

print(f"\n--- Tabuada do {numero} ---")
for i in range(1, 11):
    resultado = numero * i
    print(f"{numero} x {i:2d} = {resultado:2d}")


--- Tabuada do 7 ---
7 x  1 =  7
7 x  2 = 14
7 x  3 = 21
7 x  4 = 28
7 x  5 = 35
7 x  6 = 42
7 x  7 = 49
7 x  8 = 56
7 x  9 = 63
7 x 10 = 70


*Dica: O formato `{:2d}` dentro da f-string diz para o Python reservar 2 espaços na hora de imprimir aquele número. Isso ajuda a deixar os números alinhados bonitinhos na tela!*

## Exemplo Resolvido 2: Somatória com While

**O problema:** Peça números ao usuário repetidamente. O programa deve somar todos os números digitados e só deve parar (e mostrar o total) quando o usuário digitar o **número ZERO**.

In [ ]:
soma = 0
while True:
    valor = float(input("Digite um valor para somar (ou 0 para parar): "))

    if valor == 0:
        break # Sai do loop quando o usuário digitar zero

    soma = soma + valor

print(f"\nA soma total foi de: {soma}")

---
## Lista de Exercícios

Resolva os próximos desafios. Caso entre em loop infinito, não se esqueça de usar o botão de "Stop" ou reiniciar o Kernel.

### Desafio 1: Somatória dos Pares
Peça um número `N` ao usuário. Use um loop `for` ou `while` para calcular e exibir a SOMA de todos os números PARES entre 1 e `N`.

In [ ]:
N = int(input("Digite o valor final N: "))
soma = 0
i = 1
while i <= N:
    if i % 2 == 0:
        soma += i
    i += 1

print("Soma dos pares:", soma)

Digite o valor final N: 10
Soma dos pares: 30


### Desafio 2: Validação de Nota
Faça um programa que peça para o usuário digitar uma nota entre 0 e 10.
Sempre que o usuário digitar um valor inválido (menor que 0 ou maior que 10), o programa deve imprimir "Nota Inválida" e pedir para ele digitar a nota novamente. Só deve aceitar e encerrar o programa quando a nota for válida.
(Dica: Use um `while` para continuar pedindo até acertar)

In [ ]:
# Seu código aqui
while True:
  nota = int(input("digite uma nota entre 0 e 10: "))
  if nota < 0 or nota > 10:
    print("Nota inválida")
  else:
    break





digite um número entre 0 e 10: 12
nota inválida, tente novamente
digite um número entre 0 e 10: 4
nota válida


### Desafio 3.
Supondo que a população de um país A seja da ordem de 80000 habitantes com uma taxa anual de crescimento de 3% e que a população de B seja 200000 habitantes com uma taxa de crescimento de 1.5%. Faça um programa que calcule e escreva o número de anos necessários para que a população do país A ultrapasse ou iguale a população do país B, mantidas as taxas de crescimento.

In [ ]:
# Escreva seu código aqui
pa = 80000
pb = 200000
anos = 0
while pa < pb:
  pa *= 1.03
  pb *= 1.015
  anos += 1
print(f"serão necessários {anos} anos")

print(round(pa,2))
print(round(pb,2))

serão necessários 63 anos
515033.1
510964.42


**Desafio 4.**


Altere o programa anterior permitindo ao usuário informar as populações e as taxas de crescimento iniciais. Valide a entrada e permita repetir a operação.

In [ ]:
# Escreva seu código aqui
pa = int(input("População A: "))
pb = int(input("População B: "))
anos = 0
while pa < pb:
  pa = pa * 1.03
  pb = pb * 1.015
  anos = anos + 1
print(f"serão necessário {anos} anos")


População A: 10000
População B: 400000
serão necessário 252 anos


### Desafio 5.
Faça um programa que imprima na tela os números de 1 a 20, um abaixo do outro. Depois modifique o programa para que ele mostre os números um ao lado do outro.

In [ ]:
# Escreva seu código aqui
for x in range (1,21):
  print(x)
print("-------------------")
for x in range (1,21):
  print(x,end=" ")

1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
-------------------
1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 

### Desafio 6.
Faça um programa que leia 5 números e informe o maior número.

In [ ]:
# Escreva seu código aqui
for x in range(5):
  num = int(input("digite 5 números: "))
  if x == 0:
    maior = num
  if num > maior:
    maior = num




print(f"o maior número é: {maior}")

digite 5 números: -10
digite 5 números: 4
digite 5 números: 61
digite 5 números: 44
digite 5 números: 120
o maior número é: 120


### Desafio 7.
Faça um programa que leia 1 número inteiro e imprima o seu sucessor e o seu antecessor.

In [ ]:
# Escreva seu código aqui
numero = int(input("digite um número: "))
antecessor = numero - 1
sucessor = numero + 1
print(f"o sucessor de {numero} é {sucessor} e o antecessor é {antecessor}")

digite um número: 4
o sucessor de 4 é 5 e o antecessor é 3


### Desafio 8.
Faça um programa que leia 5 números e informe a soma e a média dos números.

In [ ]:
# Escreva seu código aqui
soma = 0
for x in range(1,6):
  num = int(input("digite 5 números: "))
  soma = soma + num
  media = soma / 5


print(f"a soma é: {soma}")
print(f"a média é: {media}")


digite 5 números: 4
digite 5 números: 12
digite 5 números: 5
digite 5 números: 24
digite 5 números: 5
a soma é: 50
a média é: 10.0


### Desafio 9.
Faça um programa que imprima na tela apenas os números ímpares entre 1 e 50.

In [ ]:
# Escreva seu código aqui
for x in range (1,51):
  if x % 2 != 0:
    print(x)

1
3
5
7
9
11
13
15
17
19
21
23
25
27
29
31
33
35
37
39
41
43
45
47
49


### Desafio 10.
Faça um programa que receba dois números inteiros e gere os números inteiros que estão no intervalo compreendido por eles.

In [ ]:
#
num1 = int(input("digite o primeiro número: "))
num2 = int(input("digite o segundo número: "))
print(f"o intervalo entre {num1} e {num2} é: ")
for x in range(num1, num2 + 1):
  print(x)



digite o primeiro número: 5
digite o segundo número: 10
o intervalo entre 5 e 10 é: 
5
6
7
8
9
10


### Desafio 11.
Altere o programa anterior para mostrar no final a soma dos números.

In [ ]:
# Escreva seu código aqui
num1 = int(input("digite o primeiro número: "))
num2 = int(input("digite o segundo número: "))
soma = num1 + num2
print(f"a soma entre {num1} e {num2} é: {soma} ")


digite o primeiro número: 51
digite o segundo número: 49
a soma entre 51 e 49 é: 100 


### Desafio 12.
Desenvolva um gerador de tabuada, capaz de gerar a tabuada de qualquer número inteiro entre 1 a 10. O usuário deve informar de qual numero ele deseja ver a tabuada. A saída deve ser conforme o exemplo abaixo:

Tabuada de 5:

5 X 1 = 5
5 X 2 = 10
...

5 X 10 = 50

In [ ]:
# Escreva seu código aqui
numero = int(input("Você quer a tabuada de qual número? "))
print(f"-------tabuada do {numero}--------")
for x in range(1,11):
  print(f"{numero} x {x} = {numero * x}")

Você quer a tabuada de qual número? 5
-------tabuada do 5--------
5 x 1 = 5
5 x 2 = 10
5 x 3 = 15
5 x 4 = 20
5 x 5 = 25
5 x 6 = 30
5 x 7 = 35
5 x 8 = 40
5 x 9 = 45
5 x 10 = 50


### Desafio 13.
Faça um programa que peça dois números, base e expoente, calcule e mostre o primeiro número elevado ao segundo número. Não utilize a função de potência da linguagem.

In [ ]:
# Escreva seu código aqui
base = int(input("digite a base do número: "))
expoente = int(input("digite o expoente: "))
elevado = base ** expoente
print(f"{base} elevado a {expoente} é igual a: {elevado}")

digite a base do número: 2
digite o expoente: 5
2 elevado a 5 é igual a: 32


### Desafio 14.
Faça um programa que peça 10 números inteiros, calcule e mostre a quantidade de números pares e a quantidade de números impares.

In [ ]:
# Escreva seu código aqui
impar = 0
par = 0
for x in range(10):
  num = int(input("digite 1 número: "))
  if num % 2 == 0:
    par += 1
  if num % 2 != 0:
    impar += 1
print(f"são {par} pares e {impar} ímpares")



digite 1 número: 1
digite 1 número: 2
digite 1 número: 3
digite 1 número: 4
digite 1 número: 5
digite 1 número: 6
digite 1 número: 7
digite 1 número: 8
digite 1 número: 9
digite 1 número: 10
são 5 pares e 5 ímpares


### Desafio 15.
A série de Fibonacci é formada pela seqüência 1,1,2,3,5,8,13,21,34,55,... Faça um programa capaz de gerar a série até o n−ésimo termo.

In [ ]:
n = int(input("qual termo você quer? "))
a = 1
b = 1
c = 0
print(a)
print(b)
for x in range(1 , n + 1):
  c = a + b
  a = b
  b = c
  print(c)





qual termo você quer? 10
1
1
2
3
5
8
13
21
34
55
89
144


### Desafio 16.
A série de Fibonacci é formada pela seqüência 0,1,1,2,3,5,8,13,21,34,55,... Faça um programa que gere a série até que o valor seja maior que 500.

In [ ]:
# Escreva seu código aqui
a = 1
b = 1
c = 0
print(a)
print(b)
while c < 500:
  c = a + b
  a = b
  b = c
  print(c)


1
1
2
3
5
8
13
21
34
55
89
144
233
377
610


### Desafio 17.
Faça um programa que calcule o fatorial de um número inteiro fornecido pelo usuário. Ex.: 5!=5.4.3.2.1=120

In [ ]:
# Escreva seu código aqui
n = int(input("Digite um número: "))
fat = 1
for x in range(1, n + 1):
  fat = fat * x
print(f"o fatorial de {n} é igual a {fat}")




Digite um número: 5
o fatorial de 5 é igual a 120


### Desafio 18.
Faça um programa que, dado um conjunto de N números, determine o menor valor, o maior valor e a soma dos valores.

In [ ]:
# Escreva seu código aqui
n = int(input("quantos números você quer? "))
i = 1
mn = 0
mo = 0
soma = 0
while i <= n:
  a = int(input("digite os números: "))
  if i == 1:
    mn = a
    mo = a
  else:
      if a < mn:
        mn = a
      if a > mo:
          mo = a
  i += 1

  soma = soma + a




print(f"maior {mo}")
print(f"menor{mn}")
print(f"a soma é {soma}")






quantos números você quer? 5
digite os números: 1
digite os números: 2
digite os números: 3
digite os números: 4
digite os números: 5
maior 5
menor1
15


### Desafio 19.
Altere o programa anterior para que ele aceite apenas números entre 0 e 1000.

In [ ]:
# Escreva seu código aqui
n = int(input('quantos números você quer? '))
maior = 0
menor = 0
soma = 0
for x in range(1, n + 1):
  p = int(input(f'digite o {x}ª número: '))
  soma = soma + p
  if p > 1000:
    print('número inválido, tente novamente')
    break
  if x == 1:
    maior = p
    menor = p
  else:
    if p > maior:
      maior = p
    if p < menor:
      menor = p

print(f'o maior número é {maior}')
print(f'o menor número é {menor}')
print(f'a soma de todos os números resulta em {soma}')


quantos números você quer? 4
digite o 1ª número: 2
digite o 2ª número: 4
digite o 3ª número: 6
digite o 4ª número: 8
o maior número é 8
o menor número é 2
a soma de todos os números resulta em 20


### Desafio 20.
Altere o programa de cálculo do fatorial, permitindo ao usuário calcular o fatorial várias vezes e limitando o fatorial a números inteiros positivos e menores que 16.

In [ ]:
while True:
    n = int(input('digite um número entre 1 e 16: '))
    if 1 <= n <= 16:
        break
    else:
        print('Número inválido, tente novamente. O número deve estar entre 1 e 16.')

fatorial = 1
for x in range(1, n + 1):
  fatorial = fatorial * x

print(f'O fatorial de {n} é igual a {fatorial}')

digite um número entre 1 e 16: 17
Número inválido, tente novamente. O número deve estar entre 1 e 16.
digite um número entre 1 e 16: 5
O fatorial de 5 é igual a 120


### Desafio 21.
Faça um programa que peça um número inteiro e determine se ele é ou não um número primo. Um número primo é aquele que é divisível somente por ele mesmo e por 1.

In [ ]:
# Escreva seu código aqui
num = int(input("digite um número inteiro: "))
divisores = 0
for x in range(1, num + 1):
  if num % x == 0:
    divisores += 1
if divisores == 2:
    print(f"{num} é primo")
else:
    print(f"{num} não é primo")




digite um número inteiro: 2
2 é primo


### Desafio 22.
Altere o programa de cálculo dos números primos, informando, caso o número não seja primo, por quais número ele é divisível.

In [ ]:
tot = 0
divisores = 0
n = int(input('digite um número inteiro: '))
for x in range(1, n + 1):
  if n % x == 0:
    tot = tot + 1
    divisores = divisores + x
print(f'o número {n} foi dividivo {tot} vezes, por {}')
if tot == 2:
  print(f'o número {n} é primo')
else:
 print(f'o número {n} não é primo')
print(divisores)


digite um número inteiro: 5
o número 5 foi dividivo 2 vezes, por 6
o número 5 é primo
6


### Desafio 23.
Faça um programa que mostre todos os primos entre 1 e N sendo N um número inteiro fornecido pelo usuário. O programa deverá mostrar também o número de divisões que ele executou para encontrar os números primos. Serão avaliados o funcionamento, o estilo e o número de testes (divisões) executados.

In [ ]:
# Escreva seu código aqui
n = int(input('quantos números você quer? '))
d = 0
for x in range(1, n + 1):
  if x % n == 0:
    d += 1
    print('é primo')
  else:
    print('não é primo')

quantos números você quer? 5
não é primo
não é primo
não é primo
não é primo
é primo


### Desafio 24.
Faça um programa que calcule o mostre a média aritmética de N notas.

In [ ]:
# Escreva seu código aqui
m = int(input('digite quantas notas você quer: '))
soma = 0
for x in range(1, m + 1):
  n = float(input(f'digite a {x}ª nota: '))
  soma = soma + n
print(soma / m)


digite quantas notas você quer: 4
digite a 1ª nota: 6
digite a 2ª nota: 7
digite a 3ª nota: 2
digite a 4ª nota: 4
4.75


### Desafio 25.
Faça um programa que peça para n pessoas a sua idade, ao final o programa devera verificar se a média de idade da turma varia entre 0 e 25,26 e 60 e maior que 60; e então, dizer se a turma é jovem, adulta ou idosa, conforme a média calculada.

In [ ]:
i = 1
media = 0
cont = 0
soma = 0
while i != 0:
  i = int(input('digite sua idade[DIGITE 0 PARA PARAR] '))
  cont = cont + 1
  soma = soma + i

media = soma / cont
print(f'a média resulta em {round(media, 2 )}')
if 0 < media <= 25:
  print('sendo classificada como jovem')
if 26 < media <= 60:
  print('sendo classificada como adulta')
if media > 61:
  print('sendo classificada como idosa')


digite sua idade[DIGITE 0 PARA PARAR] 100
digite sua idade[DIGITE 0 PARA PARAR] 101
digite sua idade[DIGITE 0 PARA PARAR] 103
digite sua idade[DIGITE 0 PARA PARAR] 104
digite sua idade[DIGITE 0 PARA PARAR] 93
digite sua idade[DIGITE 0 PARA PARAR] 97
digite sua idade[DIGITE 0 PARA PARAR] 92
digite sua idade[DIGITE 0 PARA PARAR] 12
digite sua idade[DIGITE 0 PARA PARAR] 0
a média resulta em 78.0
sendo classificada como idosa


### Desafio 26.
Numa eleição existem três candidatos. Faça um programa que peça o número total de eleitores. Peça para cada eleitor votar e ao final mostrar o número de votos de cada candidato.

In [ ]:
print('-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=')
print('ELEIÇÕES - VOTE EM SEU CANDIDATO')
print('-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=')
t1 = 0
t2 = 0
t3 = 0
ganhador = 0
e = int(input('quantas pessoas vão votar? '))
for x in range(1, e + 1):
  voto = int(input(f'faça o {x}ª voto: '))
  if voto == 1:
    t1 = t1 + 1
  elif voto == 2:
    t2 = t2 + 1
  elif voto == 3:
    t3 = t3 + 1

if t1 > t2 and t1 > t3:
  ganhador = 'Candidato 1'
elif t2 > t1 and t2 > t3:
  ganhador = 'Candidato 2'
elif t3 > t1 and t3 > t2:
  ganhador = 'Candidato 3'

print(f'foram totalizados {e} votos')
print(f'o 1ª candidato recebeu {t1} votos')
print(f'o 2ª candidato recebeu {t2} votos')
print(f'o 3ª candidato recebeu {t3} votos')
print(f'o ganhador foi {ganhador}')

-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
ELEIÇÕES - VOTE EM SEU CANDIDATO
-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
quantas pessoas vão votar? 4
faça o 1ª voto: 1
faça o 2ª voto: 1
faça o 3ª voto: 2
faça o 4ª voto: 3
foram totalizados 4 votos
o 1ª candidato recebeu 2 votos
o 2ª candidato recebeu 1 votos
o 3ª candidato recebeu 1 votos
o ganhador foi Candidato 1


### Desafio 27.
Faça um programa que calcule o número médio de alunos por turma. Para isto, peça a quantidade de turmas e a quantidade de alunos para cada turma. As turmas não podem ter mais de 40 alunos.

In [ ]:

t = int(input('quantas turmas existem? '))
soma = 0
for x in range (1, t + 1):
  a = int(input(f'quantos alunos existem na {x}ª turma? '))
  while a > 40:
      print('Não é possível a existência de mais de 40 alunos por turma')
      a = int(input(f'quantos alunos existem a {x}ª turma? '))
  soma = soma + a

  media = soma / x
print(f'a média de alunos existentes em {t} turmas é igual a {round(media, 2)}')


quantas turmas existem? 3
quantos alunos existem a 1ª turma? 41
Não é possível a existência de mais de 40 alunos por turma
quantos alunos existem a 1ª turma? 12
quantos alunos existem a 2ª turma? 31
quantos alunos existem a 3ª turma? 24
a média de alunos existentes em 3 turmas é igual a 22.33
